src/lib.rs

```diff
@@ -11,6 +11,898 @@
+                        let err = UStatus::fail_with_code(UCode::ALREADY_EXISTS, err_msg);
+                        Self::return_oneshot_result(Err(err), return_channel).await;
+                    }
+                }
+            }
+            trace!("Hit bottom of event loop");
+        }
+    }
+
+    async fn register_listener_internal(
+        source_filter: UUri,
+        sink_filter: Option<UUri>,
+        registration_type: RegistrationType,
+        msg_handler: MessageHandlerFnPtr,
+        application_wrapper: &mut UniquePtr<ApplicationWrapper>,
+        _runtime_wrapper: &UniquePtr<RuntimeWrapper>,
+    ) -> Result<(), UStatus> {
+        trace!(
+            "{}:{} - Attempting to register: source_filter: {:?} & sink_filter: {:?}",
+            UP_CLIENT_VSOMEIP_TAG,
+            UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL,
+            source_filter,
+            sink_filter
+        );
+
+        match registration_type {
+            RegistrationType::Publish(_) => {
+                trace!(
+                    "{}:{} - Registering for Publish style messages.",
+                    UP_CLIENT_VSOMEIP_TAG,
+                    UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL,
+                );
+                let (_, service_id) = split_u32_to_u16(source_filter.ue_id);
+                // let instance_id = vsomeip::ANY_INSTANCE; // TODO: Set this to 1? To ANY_INSTANCE?
+                let instance_id = 1;
+                let (_, event_id) = split_u32_to_u16(source_filter.resource_id);
+
+                trace!(
+                    "{}:{} - register_message_handler: service: {} instance: {} method: {}",
+                    UP_CLIENT_VSOMEIP_TAG,
+                    UP_CLIENT_VSOMEIP_FN_TAG_REGISTER_LISTENER_INTERNAL,
+                    service_id,
+                    instance_id,
+                    event_id
+                );
+
+                {
+                    let requested_events = REQUESTED_EVENTS.write().await;
+                    if !requested_events.contains(&(service_id, instance_id, event_id)) {
+                        get_pinned_application(application_wrapper).request_service(
+                            service_id,
+                            instance_id,
+                            ANY_MAJOR,
+                            vsomeip::ANY_MINOR,
+                        );
+                        request_single_event_safe(
+                            application_wrapper,
+                            service_id,
+                            instance_id,
+                            event_id,
+                            event_id,
+                        );
+                        get_pinned_application(application_wrapper).subscribe(
+                            service_id,
+                            instance_id,
+                            event_id,
+                            ANY_MAJOR,
+                            event_id,
```

Comment:
NOTE: specifying optional `event_id` here only filters incoming someip events from the eventgroup and invokes the callback only for event_id. Not an issue if you have 1:1 mapping.